Modified ApiWithoutInput to account for callbacks

jadahiya-MSFT opened the pull request:
Modified the onClick property to contain to separate function handlers, withPromise and withCallback. withPromise will function as it currently does, while withCallbacks will be used by API's that are run in backwards compatibility mode

erinha commented:
<p>Analyzed commit id : 04540ffeac40a43eca122f686ebf3402c4b9f97a</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: fab455cb346aff158f9ca0de677cc8355d21b421</p>

MarcinJuraszek commented:
It would be great to also move at least one API to this new model as part of this change, to show that it actually works :)

jadahiya-MSFT commented:
> It would be great to also move at least one API to this new model as part of this change, to show that it actually works :)

Added in the getContext API change and tested that it worked. Added it to this PR.

erinha commented:
<p>Analyzed commit id : f36d2f11e22774d0b27c2394155bad69d8348c5d</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: fab455cb346aff158f9ca0de677cc8355d21b421</p>

MarcinJuraszek commented on apps/teams-test-app/src/components/AppAPIs.tsx:
```diff
@@ -7,9 +7,18 @@ const GetContext = (): ReactElement =>
+          setResult(JSON.stringify(context));
+        };
+        const context = getContext(callback);
+        return JSON.stringify(context);
```
In the callback version `getContext` doesn't really return anything, does it?

```
export function getContext(callback: (context: Context) => void): void
```

so I'm not exactly sure what you're printing out out here. If I remember correctly before we moved to promises we used to initially write out `return 'API_NAME_HERE' + noHostSdkMsg;`, so in this case I think you should return `return 'getContext()' + noHostSdkMsg;` and let the callback update it to the actual values when invoked.

jadahiya-MSFT replied:
Sorry I didn't know that, I will make that change now

Ella-ly commented on apps/teams-test-app/src/components/utils/ApiWithoutInput.tsx:
```diff
@@ -18,7 +24,17 @@ export const ApiWithoutInput = (props: ApiWithoutInputProps): React.ReactElement
         value={title}
-        onClick={async () => setResult(await onClick(setResult))}
+        onClick={async () => {
+          if (typeof onClick == 'function') {
```
Can I know if you think a triple equality with '===' is better here?
` if (typeof onClick === 'function')`

jadahiya-MSFT replied:
This was just a typo, good catch! I'll update it

erinha commented:
<p>Analyzed commit id : 3442353e881449aec12dbaff6c73c5aa01a403fd</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: fab455cb346aff158f9ca0de677cc8355d21b421</p>

erinha commented:
<p>Analyzed commit id : da9d7dbe8b20e644de20b5c11c68f768c299d7fb</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>59.42 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: bd112e44bd54dff4b1b231e9709ce6490a1b5076</p>

erinha commented:
<p>Analyzed commit id : 6b6cc7141103508cf4e35d9179bea395253a9654</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.38 KB</td><td>59.37 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.38 KB</td><td>59.37 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 28b2a9862a5855e49cdfb6c3f699a02b5343f1b6</p>

TrevorJoelHarris reviewed (approved):
:shipit:

erinha commented:
<p>Analyzed commit id : dd52af0822815355429aa5d7d1f8e70f6856f532</p><hr><details><summary><b><span style="color: green">⯆</span> @microsoft/bundle-analysis-app:</b> -7 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.38 KB</td><td>59.37 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr><tr><td>Total Size</td><td>59.38 KB</td><td>59.37 KB</td><td><span style="color: green">⯆</span> -7 Bytes</td></tr></table></details><hr><p>Baseline commit: 3e0cf5a92741c9cb9c511b2231844c3f3045436b</p>
